### PR TITLE
core(config): make module resolution async

### PR DESCRIPTION
**Pull request by connorjclark**
In preparation for converting `lighthouse-core` to esm (ref #12689), the config module resolution functions need to become async to accommodate dynamic imports. This PR made `requireWrapper` async, then made every calling function async.

### CODE REVIEW

File: lighthouse-core/config/config.js [hunk 1]
```diff
@@ -148,17 +148,15 @@ function assertValidFlags(flags) {
...
  * @implements {LH.Config.Config}
  */
 class Config {
   /**
-   * @constructor
    * @param {LH.Config.Json=} configJSON
    * @param {LH.Flags=} flags
    */
-  constructor(configJSON, flags) {
+  static async createConfigFromJson(configJSON, flags) {
```
Comment (connorjclark) [inline lighthouse-core/config/config.js:159]: Constructors can't be async, so this helper function exists now.

maybe rename to `Config.createFromJson` ?

File: tsconfig-base.json [hunk 2]
```diff
@@ -7,7 +7,7 @@
     "declarationMap": true,
 
     "target": "es2020",
-    "module": "es2020",
+    "module": "es2022",
```
Comment (connorjclark) [inline tsconfig-base.json:10]: oops, i used top level await in one place... let's get #13964 done first.

File: lighthouse-core/scripts/print-a11y-scoring.js [hunk 3]
```diff
@@ -7,9 +7,8 @@
 // node lighthouse-core/scripts/print-a11y-scoring.js
 
 import Config from '../config/config.js';
-import defaultConfig from '../config/default-config.js';
 
-const config = new Config(defaultConfig);
+const config = await Config.createConfigFromJson();
```
Comment (connorjclark) [inline lighthouse-core/scripts/print-a11y-scoring.js:11]: The defaultConfig is used by default.

File: lighthouse-core/index.js [hunk 4]
```diff
@@ -83,10 +83,10 @@ async function legacyNavigation(url, flags = {}, configJSON, userConnection) {
  *   not present, the default config is used.
  * @param {LH.Flags=} flags Optional settings for the Lighthouse run. If present,
  *   they will override any settings in the config.
- * @return {Config}
+ * @return {Promise<Config>}
  */
 function generateConfig(configJson, flags) {
```
Comment (adamraine) [inline lighthouse-core/index.js:88]: TODO for me: Make this FR config and add a `generateLegacyConfig` or something.

File: lighthouse-core/test/config/config-helpers-test.js [hunk 5]
```diff
@@ -334,78 +334,78 @@ describe('.resolveSettings', () => {
   });
 });
 
-describe('.resolveGathererToDefn', () => {
+describe('.await resolveGathererToDefn', () => {
```
Comment (adamraine) [inline lighthouse-core/test/config/config-helpers-test.js:337]:
```suggestion
describe('.resolveGathererToDefn', () => {
```

**Review by adamraine — APPROVED** [review]

**Review by brendankenny — COMMENTED** [review]
Far reaching but much cleaner than I was fearing. Nice one @connorjclark!

File: lighthouse-core/config/config.js [hunk 6]
```diff
@@ -148,17 +148,15 @@ function assertValidFlags(flags) {
...
  * @implements {LH.Config.Config}
  */
 class Config {
   /**
-   * @constructor
    * @param {LH.Config.Json=} configJSON
    * @param {LH.Flags=} flags
    */
-  constructor(configJSON, flags) {
+  static async fromJson(configJSON, flags) {
```
[the remark this answers is not in the document]
Comment (adamraine) [reply lighthouse-core/config/config.js:159]: I like `initialize()` but this code is deprecated now so 🤷

[the remark this answers is not in the document]
Comment (connorjclark) [reply lighthouse-core/config/config.js:159]: The static function isn't deprecated, just the constructor. Outside of our tests there really isn't a reason you'd want to use the constructor (with this PR).

`.fromX` is a pattern that I know is common in dart (which supports multiple constructors), I wonder if that's also a pattern in other languages. so I'm leaning towards `from`. If we had unlimited characters `fromUnresolved` would be nice. but just `from` is ok too (and I'll add a note about the resolving bit in the comment)

File: lighthouse-core/scripts/print-a11y-scoring.js (hunk 3 above)
Comment (connorjclark) [reply lighthouse-core/scripts/print-a11y-scoring.js]: image

File: lighthouse-core/config/config.js (hunk 1 above)
Comment (rayjenn) [reply lighthouse-core/config/config.js]: where are my functions sir?

**Pull request merged**